### PR TITLE
Resetting cursor/hover on clear of DataTable

### PR DESCRIPTION
**Pull request by darrenburns**
## Changes

- `[Row|Cell|Column]Highlighted` event will now be emitted when the table gets its first row, assuming the cursor becomes visible (i.e. not hidden and a visible cursor_type).
- Adds considerable testing around emission of events from the DataTable.
- Adds testing around cursor position handling and clearing.
- More defensive handling of hover/cursor coordinates to account for table being cleared (I think there was a potential crash with cell cursor on clear).
- Reset the cursor positions when the table is cleared.
- Fixes edge case - if you change the cursor_type from "none" to a visible type, but `show_cursor` is False, then a highlighted event is no longer raised since the cursor still isn't visible.

**Please review the following checklist.**

- [x] Docstrings on all new or modified functions / classes 
- [x] Updated documentation
- [x] Updated CHANGELOG.md (where appropriate)

### CODE REVIEW

File: src/textual/widgets/_data_table.py [hunk 1]
```diff
@@ -14,23 +14,27 @@
...
 from ..scroll_view import ScrollView
 from ..strip import Strip
-from .._typing import Literal
 
 CursorType = Literal["cell", "row", "column", "none"]
 CELL: CursorType = "cell"
 CellType = TypeVar("CellType")
 
 
+class CellDoesntExist(Exception):
```
Comment (rodrigogiraoserrao) [inline src/textual/widgets/_data_table.py:34]:
```suggestion
class CellDoesNotExist(Exception):
```

For some reason, feels wrong to use a contraction in a class name.

File: src/textual/widgets/_data_table.py [hunk 2]
```diff
@@ -229,9 +233,16 @@ def get_cell_value(self, coordinate: Coordinate) -> CellType:
...
+
+        Raises:
+            CellDoesntExist: If there is no cell with the given coordinate.
         """
         row, column = coordinate
-        return self.data[row][column]
+        try:
+            cell_value = self.data[row][column]
+        except KeyError:
+            raise CellDoesntExist(f"No cell exists at {coordinate!r}")
```
Comment (rodrigogiraoserrao) [inline src/textual/widgets/_data_table.py:244]: This feels like the perfect place for you to teach me this: won't this raise the `CellDoesn'tExist` and show a traceback saying something "KeyError was raised and during handling of that exception, `CellDoesn'tExist` was raised"? Don't we need to add a `from None` here?

File: src/textual/widgets/_data_table.py [hunk 3]
```diff
@@ -293,18 +304,26 @@ def watch_cursor_cell(
...
+            return
+        else:
+            self.emit_no_wait(DataTable.CellHighlighted(self, cell_value, coordinate))
 
     def _highlight_row(self, row_index: int) -> None:
         """Apply highlighting to the row at the given index, and emit event."""
         self.refresh_row(row_index)
-        self.emit_no_wait(DataTable.RowHighlighted(self, row_index))
+        if row_index in self.data:
+            self.emit_no_wait(DataTable.RowHighlighted(self, row_index))
```
Comment (rodrigogiraoserrao) [inline src/textual/widgets/_data_table.py:320]: `self.refresh_row` will also check that we are trying to refresh a row that exists, so is that why you don't have the call to `refresh_row` inside the if statement?

**Review by rodrigogiraoserrao — APPROVED** [review]
Looks good to me, although I did leave a couple of questions!

File: src/textual/widgets/_data_table.py (hunk 3 above)
Comment (darrenburns) [reply src/textual/widgets/_data_table.py:320]: Yep, that's correct.

File: src/textual/widgets/_data_table.py (hunk 1 above)
Comment (darrenburns) [reply src/textual/widgets/_data_table.py:34]: Yeah, I think this is better. I'll update this.

File: src/textual/widgets/_data_table.py (hunk 2 above)
Comment (davep) [reply src/textual/widgets/_data_table.py]: I'd second a `from None` here.

Comment (darrenburns) [reply src/textual/widgets/_data_table.py]: Added `from None`

**Review by willmcgugan — APPROVED** [review]

**Pull request merged**